<a href="https://colab.research.google.com/github/sofiavillegas07/Anal-tica/blob/main/Reto_2_%C3%81rbol_de_decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 2: Árboles de decisión**
0. Cargar las librerias de referencia para el modelo


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Carga de los datos del archivo travel data

In [ ]:
nxl='/content/drive/MyDrive/Analitica de negocios /2. BD1_Travel_Data_int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0) #No me funcionó con sheet_name=1
XDB=XDB.dropna()
XDB.head(100)

XD=np.array(XDB[['College','FoodSpend','Income','TravelSpend']])
print(XD)
yd=np.array(XDB['CreditCard']) #El dato que quiero que el modelo pronostique es si el cliente va a utilizar tarjeta de crédito
print(yd)


[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[1 1 1 1 1 0 0 1 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1 0 1 0 1 1 0 1 0 1 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 1
 0 0 1 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 1 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 0 0 0
 1 0 1 0 0 1 1 1 0 0 0 1 1 0 0 1 0 0 0 0 1 0 0 0 0 0 1 0 1 1 0 0 0 1 0 0 0
 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 1 1 0 0 0 1 0 0 1 0 0 0 0 0
 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 0 1 0 0 1 0 1 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0
 0 0 0 0 0 0 1 1 1 1 0 1 1 1 0 1 1 1 1 1 0 0 0 1 0 0 0 0 1 1 1 0 0 1 0 1 0
 0 0 0 0 0 1 

2. Creación del modelo - Decision Tree

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4) #Dividido en 4 niveles
mar.fit(XD,yd)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

#Evaluamos el modelo
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Ex=(VN+VP)/(VN+VP+FN+FP) #Que porcentaje de los datos pronostico correctamente
print("La exactitud alcanzada por el modelo es:", Ex)
Se=VP/(FN+VP) #Que porcentaje de los datos positivos pronosticado correctamente
print("La sensibilidad alcanzada por el modelo es:", Se)
Es=VN/(VN+FP) #Que porcentaje de los datos positivos pronosticado correctamente
print("La especificidad alcanzada por el modelo es:", Es)

[[334   9]
 [ 42 115]]
La exactitud alcanzada por el modelo es: 0.898
La sensibilidad alcanzada por el modelo es: 0.732484076433121
La especificidad alcanzada por el modelo es: 0.9737609329446064


3. Graficar el árbol

In [ ]:
#Estas son las librerias para graficar el arbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College','FoodSpend','Income','TravelSpend'] #son los titulos para los nodos
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol.png')

True

4. Pronostico de las personas que van a pagar con tarjeta de crédito

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2=XDB2.dropna()
XDB2.head(100)

XD2=np.array(XDB2[['College','FoodSpend','Income','TravelSpend']]) #Datos de entrada
print(XD2)

#Pronosticamos las personas que van a pagar con tarjeta
ydp2=mar.predict(XD2)
print(ydp2)



[[0.00000e+00 2.89290e+03 6.59820e+04 2.15160e+03]
 [1.00000e+00 6.01766e+03 5.39860e+04 2.00250e+03]]
[0 0]


**Analisis**
De acuerdo con la base de datos suministrada y el nodo base del arbol de decision, se puede observar un total de 500 datos.

De acuerdo con el pronostico realizado por el modelo, este alcanzo una exactitud cercana al 89%, valores que fueron similares a la sensibilidad que fue de 73% y a la especifidad que fue del 93%.

Por otro lado, de acuerdo con los nuevos clientes en la base de datos Travel_Plan_Score, la probabilidad de que el primer cliente utilizará la tarjeta de crédito para el pago de los gastos de viaje y de los alimentos en sus planes de viajar el próximo año es del 100%.

La probabilidad de que el segundo cliente utilizará la tarjeta de crédito para el pago de los gastos de viaje y de los alimentos en sus planes de viajar el próximo año es del 78%.

Las reglas del negocio a las que pertenecen los dos individuos anteriores:

*Primera persona:*
TravelSpend>=1567.875 AND Income>59098.0 [108-0]

*Segunda persona:*
TravelSpend>=1567.875 AND Income<59098.0 AND FoodSpend>2711.895 AND TravelSpend<=3008.25 [107-29]

*Las reglas del negocio que poseen nodos puros son:*
- TravelSpend<=1567.875 AND TravelSpend<1188.0 AND FoodSpend>4990.73 [0-47]

- TravelSpend<=1567.875 AND TravelSpend<1188.0 AND FoodSpend<4990.73  FoodSpend>4972 [1-0]

- TravelSpend>=1567.875 AND Income>59098 [108-0]





